### PR TITLE
Add initial sentry tracing extension

### PR DESCRIPTION
**Pull request by patrick91**
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

*

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
## Apollo Federation Subgraph Compatibility Results

<table>
<thead>
<tr><th>Federation 1 Support</th><th>Federation 2 Support</th></tr>
</thead>
<tbody>
<tr><td><table><tr><th><code>_service</code></th><td>🟢</td></tr><tr><th><code>@key (single)</code></th><td>🟢</td></tr><tr><th><code>@key (multi)</code></th><td>🟢</td></tr><tr><th><code>@key (composite)</code></th><td>🟢</td></tr><tr><th><code>repeatable @key</code></th><td>🟢</td></tr><tr><th><code>@requires</code></th><td>🟢</td></tr><tr><th><code>@provides</code></th><td>🟢</td></tr><tr><th><code>federated tracing</code></th><td>🔲</td></tr></table></td><td><table><tr><th><code>@link</code></th><td>🟢</td></tr><tr><th><code>@shareable</code></th><td>🟢</td></tr><tr><th><code>@tag</code></th><td>🟢</td></tr><tr><th><code>@override</code></th><td>🟢</td></tr><tr><th><code>@inaccessible</code></th><td>🟢</td></tr></table></td></tr>
</tbody>
</table>

Learn more:
* [Apollo Federation Subgraph Specification](https://www.apollographql.com/docs/federation/subgraph-spec/)
* [Compatibility Tests](https://github.com/apollographql/apollo-federation-subgraph-compatibility/blob/main/COMPATIBILITY.md)

**Comment by botberry** [comment]
# Pre-release
:wave:

Releasing commit [f192bb5ecbbd1d22faf603e566e99a435df05d9e] to PyPi as pre-release! :package:

**Comment by botberry** [comment]
# Pre-release
:wave:

Pre-release **0.156.0.dev.1674920499** [f192bb5ecbbd1d22faf603e566e99a435df05d9e] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.156.0.dev.1674920499
```

**Comment by jerber** [comment]
Is there a way to exclude certain queries / mutations from being traced?

**Comment by patrick91** [comment]
@jerber not at the moment, but we should allow it I think

**Comment by patrick91** [comment]
@jerber we should probably also consolidate all the tracing extensions :D

what's your usecase for skipping some queries? :)

**Comment by jerber** [comment]
@patrick91 We have a decorator for any function that controls the Sentry sampling rate for that function: `@Sentry.sample_rate(0.5)` which works well

File: strawberry/extensions/tracing/sentry.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        self.validation_span.finish()
+
+    def on_parsing_start(self):
+        self.parsing_span = self.gql_span.start_child(
+            op="parsing", description="Parsing"
+        )
+
+    def on_parsing_end(self):
+        self.parsing_span.finish()
+
```
Comment (Ronjea) [inline strawberry/extensions/tracing/sentry.py:82]: Hi @patrick91, what do you think adding

```
    def on_execute(self):
        with start_transaction(
                op=self._operation_type,
                name=self._operation_name,
        ):
            yield
```

With this we can also have the option to track the transaction in sentry 😊

Comment (patrick91) [reply strawberry/extensions/tracing/sentry.py:82]: @Ronjea I'll do this once I get back to this PR, thanks for the suggestion!

Comment (Ronjea) [reply strawberry/extensions/tracing/sentry.py:82]: My pleasure ☺️

Comment (patrick91) [reply strawberry/extensions/tracing/sentry.py:91]: @Ronjea can you take a look at this PR? 

I'll merge this soon, but happy to update the extension with your suggestions later on

**Comment by patrick91** [comment]
@jerber I've added a method on the extension, I think we can allow passing a callable in future, but we should it for all the tracing extensions, so I'll postpone that for now

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release adds a SentryTracingExtension that you can use to automatically add
tracing information to your GraphQL queries.


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by leo6104** [comment]
I think its `SentryTracingExtension` class is missing in PyPI released version 0.177.0 

Can you check its release version?

**Comment by patrick91** [comment]
@leo6104 does it work with: 

```
from strawberry.extensions.tracing.sentry import SentryTracingExtension
```

?

**Comment by leo6104** [comment]
Sorry for my mistake. it was my local cache problem. 😅

**Comment by aprams** [comment]
Thanks a lot @patrick91 ! Great addition!

**Pull request merged**